[BUG] 'runway tfenv install' broken

### Bug Description

Hello,

This started happening yesterday Apr 13, 2022

We've tried all sort of combinations of runway/terraform/python versions and cannot seem to make this work anymore.

This is affect our CI/CD pipeline which can no longer install the version of terraform specified in the runway.yaml file because runway, i believe, is using runway tfenv to install the desired version of terraform

Any help would be greatly appreciated. Until then i'll be creating my own "wrapper" to install the desired versions of terraform myself into the directory runway is looking for $HOME/.tfenv/versions

Thanks!

```

devops/tfetest|~/repos/innovyze/devops-terraform
😄-bash-⌲ runway tfenv install 1.0.0
Traceback (most recent call last):
  File "/Users/patrick.raco/.local/bin/runway", line 8, in <module>
    sys.exit(cli())
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/_cli/main.py", line 33, in invoke
    return super().invoke(ctx)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/_cli/commands/_tfenv/_install.py", line 31, in install
    "terraform path: %s", TFEnvManager().install(version_requested=version)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/env_mgr/tfenv.py", line 348, in install
    if not self.version:
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/backports/cached_property/__init__.py", line 80, in __get__
    val = self.func(instance)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/env_mgr/tfenv.py", line 276, in version
    for i in get_available_tf_versions(include_prerelease_versions)
  File "/Users/patrick.raco/.local/lib/python3.7/site-packages/runway/env_mgr/tfenv.py", line 118, in get_available_tf_versions
    reverse=True,
  File "/Users/patrick.raco/.pyenv/versions/3.7.12/lib/python3.7/distutils/version.py", line 52, in __lt__
    c = self._cmp(other)
  File "/Users/patrick.raco/.pyenv/versions/3.7.12/lib/python3.7/distutils/version.py", line 337, in _cmp
    if self.version < other.version:
TypeError: '<' not supported between instances of 'str' and 'int'```

### Expected Behavior

1. run: `runway tfenv install VERSION`
2. expect specified terraform version to be installed


### Steps To Reproduce

1. install `runway`: `pip install runway [--user]`
2. make sure desired version of terraform is not installed with: `runway tfenv list`
3. try to install it: `runway tfenv install --version`

### Runway version

- 1.18.1
- 1.18.3
- 2.6.1

### Installation Type

pypi (pip, pipenv, poetry, etc)

### OS / Environment

- macos
- linux

### Anything else?

- seems to happen if/when the desired version of terraform is NOT installed
- Did notice these diffs in the installed python packages between a successful and failed run:

![image](https://user-images.githubusercontent.com/3203156/163601117-349be257-6cfc-4f1b-8b97-82237ed6bc91.png)

in case a solution isn't found soon, hope these will help someone else:

```
#!/bin/bash
#
# purpose: install required terraform versions listed in .terraform-version
#

# enable bash debugging if set on the pipeline
if grep -Fiwq "true" <<< "$SYSTEM_DEBUG"; then set -x; fi

# globals
DEVOPS_SCRIPTS=$DEVOPS_ADO_SCRIPTS
TERRAFORM_VERSION_FILE=".terraform-version"

# loop through list of required terraform versions and install them
if [[ -f "$TERRAFORM_VERSION_FILE" ]]; then
  # shellcheck disable=SC2013
  for terraform_version in $(cat $TERRAFORM_VERSION_FILE); do
    echo "Installing Terraform version ($terraform_version) found in: $TERRAFORM_VERSION_FILE"
    "$DEVOPS_SCRIPTS/install-terraform.sh" "$terraform_version"
  done
else
  echo "The Terraform version file NOT found: $TERRAFORM_VERSION_FILE"
  echo "Cannot install Terraform versions needed in runway.yaml"
  exit 1
fi
```

```
#!/bin/bash
#
# purpose: install terraform version
#
# description: simple script to install a specific terraform version
# 
# note: the versions of terraform are installed into the
#       directory that runway is looking for so that
#       runway does not try to install it itself using `runway tfenv`
#       which is currently broken as of (13-APR-2022)
#

# enable bash debugging if set on the pipeline
if grep -Fiwq "true" <<< "$SYSTEM_DEBUG"; then set -x; fi

THIS_SCRIPT=$(basename "$0")
TF_ENV_DIR="$HOME/.tfenv/versions"  # this is where runway is looking
# create the directory if it doesn't exist
[[ ! -d $TF_ENV_DIR ]] && mkdir -p $TF_ENV_DIR
TF_RELEASES_URL="https://releases.hashicorp.com"
USAGE=\
"usage:
  $THIS_SCRIPT VERSION      install specified terraform version
  $THIS_SCRIPT -s | --show  show installed terraform versions
  $THIS_SCRIPT -h | --help  disploy this help/usage"

# get the version to install
tf_version_desired=$1

function show_installed_versions {
  # shellcheck disable=SC2086
  if ls -d $TF_ENV_DIR/* &> /dev/null; then
    echo "these are the installed versions:"
    ls "$TF_ENV_DIR"
  else
    echo "there are no installed versions"
  fi
}

# command: help - show help
if [[ "$tf_version_desired" == "-h" ]] || [[ "$tf_version_desired" == "--help" ]]; then
  echo "this function helps to install a specific terraform version"
  echo "$USAGE"
# command: show - show terraform versions installed/available
elif [[ "$tf_version_desired" == "-s" ]] || [[ "$tf_version_desired" == "--show" ]]; then
  show_installed_versions
# command: "install" (default) - install the specified terraform version
elif [[ -n "$tf_version_desired" ]]; then
  if [[ -x "$TF_ENV_DIR/$tf_version_desired/terraform" ]]; then
    echo "desired version ($tf_version_desired) already installed in '$TF_ENV_DIR'"
  else
    echo "did not find desired version ($tf_version_desired) in '$TF_ENV_DIR'"
    show_installed_versions
    echo "going to attempt to install it"
    host_arch=$(uname -m)
    case $host_arch in
      x86_64) arch=amd64;;
      arm64) arch=arm64;;
      *) arch=unknown;;
    esac
    host_os=$(uname -s)
    case $host_os in
      Darwin) os=darwin;;
      Linux) os=linux;;
      *) os=unknown;;
    esac
    if [[ "$arch" == "unknown" ]] && [[ "$os" == "unknown" ]]; then
      echo "cannot determine correct terraform binary to download for (Arch: $host_arch, O.S.: $host_os)"
      exit
    fi
    tf_version_zip_name="terraform_${tf_version_desired}_${os}_${arch}.zip"
    tf_version_url="${TF_RELEASES_URL}/terraform/${tf_version_desired}/${tf_version_zip_name}"
    if curl -s -f -o "/tmp/${tf_version_zip_name}" "${tf_version_url}"; then
      [[ ! -d "$TF_ENV_DIR" ]] && mkdir "$TF_ENV_DIR"
      unzip "/tmp/${tf_version_zip_name}" -d "$TF_ENV_DIR/$tf_version_desired"
      echo "successfully installed desired version ($tf_version_desired) in '$TF_ENV_DIR'"
      show_installed_versions
    else
      tf_available_versions="$(curl -s "${TF_RELEASES_URL}/terraform/" | grep -oE 'terraform_\d+\.\d+.\d+' | sort -uV | sed 's/^terraform_//' | tr '\n' ',' | sed 's/,/, /g')"
      echo "cannot find desired version ($tf_version_desired) at '$TF_RELEASES_URL'"
      echo "these are the available versions:"
      echo "   $tf_available_versions"
      exit
     fi
  fi
# command: none - show usage
else
  echo "error: incorrect usage"
  echo "$USAGE"
fi
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -49,7 +49,7 @@ repos:
     hooks:
       - id: isort
   - repo: https://github.com/psf/black
-    rev: 22.1.0
+    rev: 22.3.0
     hooks:
       - id: black
         args:

--- a/runway/cfngin/hooks/staticsite/upload_staticsite.py
+++ b/runway/cfngin/hooks/staticsite/upload_staticsite.py
@@ -300,7 +300,7 @@ def calculate_hash_of_extra_files(
     """
     file_hash = hashlib.md5()
 
-    for extra_file in sorted(extra_files, key=lambda extra_file: extra_file.name):
+    for extra_file in sorted(extra_files, key=lambda x: x.name):
         file_hash.update((extra_file.name + "\0").encode())
 
         if extra_file.content_type:

--- a/runway/env_mgr/tfenv.py
+++ b/runway/env_mgr/tfenv.py
@@ -13,7 +13,6 @@ import subprocess
 import sys
 import tempfile
 import zipfile
-from distutils.version import LooseVersion
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -31,6 +30,7 @@ from urllib.request import urlretrieve
 import hcl
 import hcl2
 import requests
+from packaging.version import LegacyVersion
 from typing_extensions import Final
 
 from ..compat import cached_property
@@ -114,7 +114,7 @@ def get_available_tf_versions(include_prerelease: bool = False) -> List[str]:
     )["terraform"]
     tf_versions = sorted(
         [k for k, _v in tf_releases["versions"].items()],  # descending
-        key=LooseVersion,
+        key=LegacyVersion,
         reverse=True,
     )
     if include_prerelease:

--- a/tests/unit/env_mgr/test_tfenv.py
+++ b/tests/unit/env_mgr/test_tfenv.py
@@ -70,8 +70,8 @@ def test_get_available_tf_versions(mocker: MockerFixture) -> None:
     mock_requests.get.return_value = MagicMock(text=json.dumps(response))
     assert get_available_tf_versions() == ["0.12.0"]
     assert get_available_tf_versions(include_prerelease=True) == [
-        "0.12.0-beta",
         "0.12.0",
+        "0.12.0-beta",
     ]
 
 

